### PR TITLE
gui: show connected miners

### PR DESCRIPTION
**Pull request by jholdstock**
### gui: Fix and simplify code.
This fixes some issues with the gui which were introduced
in the previous commit. Also:

- Combine methods in hub.go so pool stats can be fetched in one call.
- Remove truncate param from method to fetch mined work
- Remove minNano param from method to fetch payments
- Reverse slices in index.go so payments and mined blocks
are displayed in reverse order. (This is temporary)

### gui: show connected miners.
- index: show connected miners belonging to account
- move usage of util.HashString into the templates
- admin: show all connected miners
- Move AccountStats struct to gui package
- Add server side input validation to the Address form
- Combine functions FetchHash and FetchConnections so we only need to iterate
over connections once.

### CODE REVIEW

File: gui/index.go [hunk 1]
```diff
@@ -44,27 +61,54 @@ func (ui *GUI) GetIndex(w http.ResponseWriter, r *http.Request) {
...
-	data.AccountStats = resp
+	if len(work) == 0 && len(payments) == 0 {
+		log.Infof("Nothing found for address: %s", address)
+		data.Error = fmt.Sprintf("Nothing found for address: %s", address)
+		ui.renderTemplate(w, r, "index", data)
+		return
+	}
+
+	// // Reverse for display purposes. We want the most recent block/payments to be first.
+	reverseSlice(work)
```
Comment (dnldd) [inline gui/index.go]: `ListMinedWork` already returns a reversed slice, it iterates and filters from the last entry to the first.

File: gui/index.go [hunk 2]
```diff
@@ -44,27 +61,54 @@ func (ui *GUI) GetIndex(w http.ResponseWriter, r *http.Request) {
...
+	reverseSlice(payments)
```
Comment (dnldd) [inline gui/index.go]: Update `FetchArchivedPaymentsForAccount` to iterate entries in reverse instead and remove `reverseSlice`.

Use: `for k, v := c.Last(); k != nil; k, v = c.First()`

**Comment by jholdstock** [comment]
All review points have been addressed as we agreed on matrix.

- All data displayed on the GUI is now truncated to the most recent 10 entries. Temporary until we properly implement paging and can display all data efficiently
- Data is ordered during the db access using `Last` and `Prev`, rather than `First` and `Next`
- reverseSlice function removed
-

File: dividend/payment.go [hunk 3]
```diff
@@ -571,28 +576,28 @@ func FetchArchivedPaymentsForAccount(db *bolt.DB, account []byte, minNano []byte
...
+
+				if len(pmts) == int(n) {
+					return nil
+				}
 			}
 		}
 
 		return nil
 	})
-	if err != nil {
```
Comment (dnldd) [inline dividend/payment.go]: maintain this check, let's avoid heap allocation where we can.

File: network/acceptedwork.go [hunk 4]
```diff
@@ -178,26 +182,27 @@ func ListMinedWork(db *bolt.DB, truncate bool) ([]*AcceptedWork, error) {
...
-				if truncate && len(minedWork) == truncatedSize {
+				if len(minedWork) == int(n) {
 					return nil
 				}
 			}
 		}
 
 		return nil
 	})
-	if err != nil {
```
Comment (dnldd) [inline network/acceptedwork.go]: please maintain.

File: network/acceptedwork.go [hunk 5]
```diff
@@ -219,16 +224,16 @@ func ListMinedWorkByAccount(db *bolt.DB, accountID string) ([]*AcceptedWork, err
...
 				minedWork = append(minedWork, &work)
+				if len(minedWork) == int(n) {
+					return nil
+				}
 			}
 		}
 
 		return nil
 	})
-	if err != nil {
```
Comment (dnldd) [inline network/acceptedwork.go]: please maintain.

File: gui/index.go [hunk 6]
```diff
@@ -44,27 +60,42 @@ func (ui *GUI) GetIndex(w http.ResponseWriter, r *http.Request) {
 
 	data.Address = address
 
-	minedWork, err := ui.hub.FetchMinedWork(true)
+	if len(address) != 35 {
```
Comment (dnldd) [inline gui/index.go]: Sorry, missed this earlier. Please decode the address and assert it's derived from the active network set in the config. Refer to: https://github.com/decred/dcrpool/blob/master/config.go#L443-L457

Comment (jholdstock) [reply gui/index.go]: added

Comment (jholdstock) [reply gui/index.go]: added

**Pull request merged**